tests/unit/hive_metastore/test_principal_grants.py

```diff
@@ -0,0 +1,254 @@
+from unittest.mock import create_autospec
+
+import pytest
+from databricks.labs.blueprint.installation import MockInstallation
+from databricks.labs.lsql.backends import StatementExecutionBackend
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import ResourceDoesNotExist
+from databricks.sdk.service import iam
+from databricks.sdk.service.catalog import ExternalLocationInfo
+
+from databricks.labs.ucx.assessment.azure import (
+    AzureServicePrincipalCrawler,
+    AzureServicePrincipalInfo,
+    ServicePrincipalClusterMapping,
+)
+from databricks.labs.ucx.azure.access import AzureResourcePermissions
+from databricks.labs.ucx.azure.resources import AzureAPIClient, AzureResources
+from databricks.labs.ucx.config import WorkspaceConfig
+from databricks.labs.ucx.hive_metastore import ExternalLocations, Mounts, TablesCrawler
+from databricks.labs.ucx.hive_metastore.grants import Grant, PrincipalACL
+from databricks.labs.ucx.hive_metastore.locations import Mount
+from databricks.labs.ucx.hive_metastore.tables import Table
+
+
+@pytest.fixture
+def ws():
+    w = create_autospec(WorkspaceClient)
+    w.config.is_azure = True
+    w.external_locations.list.return_value = [
+        ExternalLocationInfo(url="abfss://container1@storage1.dfs.core.windows.net/folder1"),
+        ExternalLocationInfo(url="abfss://container1@storage2.dfs.core.windows.net/folder2"),
+        ExternalLocationInfo(url="abfss://container1@storage3.dfs.core.windows.net/folder3"),
+    ]
+
+    permissions = {
+        'cluster1': iam.ObjectPermissions(
+            object_id='cluster1',
+            object_type="clusters",
+            access_control_list=[
+                iam.AccessControlResponse(group_name='group1', all_permissions=[iam.Permission(inherited=False)]),
+                iam.AccessControlResponse(
+                    user_name='foo.bar@imagine.com',
+                    all_permissions=[iam.Permission(permission_level=iam.PermissionLevel.CAN_USE)],
+                ),
+            ],
+        ),
+        'cluster2': iam.ObjectPermissions(
+            object_id='cluster2',
+            object_type="clusters",
+            access_control_list=[
+                iam.AccessControlResponse(
+                    service_principal_name='spn1',
+                    all_permissions=[iam.Permission(permission_level=iam.PermissionLevel.CAN_USE)],
+                ),
+            ],
+        ),
+        'cluster3': iam.ObjectPermissions(object_id='cluster2', object_type="clusters"),
+    }
+    w.permissions.get.side_effect = lambda _, object_id: permissions[object_id]
+    return w
+
+
```

Comment:
```suggestion
def principal_acl(ws, install, cluster_spn: list):
```
Be consistent with naming